src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs

```diff
@@ -27,17 +27,18 @@ public class PartialDateTime
         private const string SecondCapture = "second";
         private const string FractionCapture = "fraction";
         private const string TimeZoneCapture = "timeZone";
+        private const string InvalidTimeZoneCapture = "invalidTimeZone";
 
-        // There is a difference between how the date should be parsed for storing and how the date should be parsed for searching.
-        // For date that should be stored, the time zone information must be present if time is specified.
-        // From spec: http://hl7.org/fhir/STU3/datatypes.html#datetime, "If hours and minutes are specified, a time zone SHALL be populated."
-        // However, if the date is being parsed for searching, then the time zone information is optional.
-        // From spec: http://hl7.org/fhir/STU3/search.html#date, "the minutes SHALL be present if an hour is present, and you SHOULD provide a time zone if the time part is present."
-        // This regular expression allows the time zone to be optional.
+        // This regular expression is used to capture which date time parts are specified by the user and which parts are not.
+        // This is required because date time parts left blank are used to indicate a time period.
+        // For example, 2000 is equivalent to an interval of [2000-01-01T00:00, 2000-12-31T23:59].
         private static readonly Regex DateTimeRegex = new Regex(
-            $@"-?(?<{YearCapture}>[0-9]{{4}})(-(?<{MonthCapture}>0[1-9]|1[0-2])(-(?<{DayCapture}>0[0-9]|[1-2][0-9]|3[0-1])(T(?<{HourCapture}>[01][0-9]|2[0-3]):(?<{MinuteCapture}>[0-5][0-9])(:((?<{SecondCapture}>[0-5][0-9])(?<{FractionCapture}>\.[0-9]+)?))?(?<{TimeZoneCapture}>Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))?)?)?)?",
+            $@"-?(?<{YearCapture}>[0-9]{{4}})(-(?<{MonthCapture}>[0-9]{{2}}))?(-(?<{DayCapture}>[0-9]{{2}}))?(T(?<{HourCapture}>[0-9]{{2}}))?(:(?<{MinuteCapture}>[0-9]{{2}}))?(:(?<{SecondCapture}>[0-9]{{2}}))?((?<{FractionCapture}>\.[0-9]+))?((?<{TimeZoneCapture}>Z|(\+|-)(([0-9]{{2}}):[0-9]{{2}}))|(?<{InvalidTimeZoneCapture}>Z|(\+|-)(([0-9]{{1}}):[0-9]{{2}})))?",
```

Comment:
You can replace the [0-9]s with \d and the {{1}} towards the end is redundant.

Reply:
Awesome, I was in need of feedback from a regex expert. Let me know how b05e9ef looks.